### PR TITLE
feat: add query param to add a `tip`, `partialFee`, and `priority` field to each extrinsic for `/node/transaction-pool`

### PR DESCRIPTION
**Pull request by TarikGul**
## Feature

**Summary**: Allow `/node/transaction-pool` to accept a query parameter called `includeFee`. When true it will attach a decoded `tip`, `partialFee`, and `priority` to each extrinsic. 

### Query Param `includeFee`

- Default: It will be `false`.
- If true: It will attach a `tip`, `partialFee`, and `priority` to each extrinsic within the "pool" array. 

rel: https://github.com/paritytech/substrate-api-sidecar/issues/925

### CODE REVIEW

File: src/controllers/node/NodeTransactionPoolController.ts [hunk 1]
```diff
@@ -29,12 +30,14 @@ export default class NodeTransactionPoolController extends AbstractController<No
 	 * @param res Express Response
```
Comment (CurlyBracketEffect) [inline src/controllers/node/NodeTransactionPoolController.ts:30]: Just a little note to adjust the comment describing the function params.

**Review by CurlyBracketEffect — APPROVED** [review]
Logic looks good...Just the note about the comment.

File: src/controllers/node/NodeTransactionPoolController.ts (hunk 1 above)
Comment (CurlyBracketEffect) [reply src/controllers/node/NodeTransactionPoolController.ts:30]: This comment was supposed to reference line 28

Comment (TarikGul) [reply src/controllers/node/NodeTransactionPoolController.ts:30]: Good catch.

**Review by lexnv — APPROVED** [review]

File: docs/src/openapi-v1.yaml [hunk 2]
```diff
@@ -2765,6 +2773,10 @@ components:
                   type: string
                   format: hex
                   description: Scale encoded extrinsic.
+                tip:
+                  type: string
+                  format: number
```
Comment (TarikGul) [inline docs/src/openapi-v1.yaml:2778]:
```suggestion
                  format: unsignedInteger
```

File: src/services/node/NodeTransactionPoolService.ts [hunk 3]
```diff
@@ -3,20 +3,32 @@ import { INodeTransactionPool } from 'src/types/responses';
...
-	async fetchTransactionPool(): Promise<INodeTransactionPool> {
+	async fetchTransactionPool(
+		includeTip: boolean
+	): Promise<INodeTransactionPool> {
 		const { api } = this;
 
 		const extrinsics = await api.rpc.author.pendingExtrinsics();
 
-		const pool = extrinsics.map((ext) => {
+		if (includeTip) {
```
Comment (jsdw) [inline src/services/node/NodeTransactionPoolService.ts:13]: Wondering whether you can avoid the duplication and get away with something like:

```ts
return {
    pool: extrinsics.map((ext) => {
        return {
            hash: ext.hash.toHex(),
            encodedExtrinsic: ext.toHex(),
            tip: includeTip ? ext.tip.toString() : undefined,
        };
    }),
};
```

(afaik undefined fields disappear when serialized to JSON anyway)

**Review by jsdw — APPROVED** [review]

**Comment by jsdw** [comment]
Just to relay @joepetrowski's comment for consideration here:

Wouldn't you want to return the total fee including tip?
Since that's what determines priority. So if a transfer has a fee of 0.1, and a remark has a fee of 0.001, and the pool you query is loaded with transfers with a tip of 0, you could then think, "I'll put a tip of 0.01 just to make sure I'm ahead of all the stingy non-tippers", but with a total fee of 0.011 you'll still be last in line.

**Comment by TarikGul** [comment]
@jsdw Yea I am going to rewrite this a bit to include the patialFee, and totalFee. Will request reviews again after.

File: src/services/node/NodeTransactionPoolService.ts [hunk 4]
```diff
@@ -14,24 +14,103 @@
...
+					'This runtime does not have the sufficient materials to calculate fees.'
+				);
+			}
+
+			return {
+				pool: extrinsics.map((ext) => {
+					const { hash, encodedLength, tip } = ext;
+					const partialFee = calcFeeWithParams.calc_fee(
+						// TODO This value comes from the success, failure events.
+						BigInt(0),
```
Comment (TarikGul) [inline src/services/node/NodeTransactionPoolService.ts:47]: This is the last piece of this calculation to figure out. Originally this value comes from the weight provided by a failed or successful event.

Comment (TarikGul) [reply src/services/node/NodeTransactionPoolService.ts:47]: Might make more sense to use `api.rpc.payment.queryInfo` to find fee estimate

Comment (TarikGul) [reply src/services/node/NodeTransactionPoolService.ts:47]: Okay so this is the right move, I am going to remove all `@substrate/calc` logic here and replace it with `queryInfo`.

File: src/services/test-helpers/mock/mockApi.ts [hunk 5]
```diff
@@ -28,8 +28,10 @@ import {
 	SessionIndex,
 	StakingLedger,
 } from '@polkadot/types/interfaces';
+import BN from 'bn.js';
 
 import { polkadotMetadata } from '../../../test-helpers/metadata/metadata';
+// import { TypeFactory } from '../../../test-helpers/typeFactory';
```
Comment (niklasad1) [inline src/services/test-helpers/mock/mockApi.ts:34]:
```suggestion
```

Comment (niklasad1) [inline src/services/test-helpers/mock/mockApi.ts:34]:
```suggestion
```

File: src/services/node/NodeTransactionPoolService.ts [hunk 6]
```diff
@@ -14,24 +14,172 @@
...
+		ext: Extrinsic,
+		dispatchClass: DispatchClass,
+		weight: Weight
+	): Promise<string> {
+		const { api } = this;
+		const { tip, encodedLength: len } = ext;
+		const BN_ONE = new BN(1);
+		const sanitizedClass = this.defineDispatchClassType(dispatchClass);
+
+		const { maxBlock } = this.api.consts.system.blockWeights;
```
Comment (niklasad1) [inline src/services/node/NodeTransactionPoolService.ts:106]:
```suggestion
		const { maxBlockWeight } = this.api.consts.system.blockWeights;
```

Comment (niklasad1) [inline src/services/node/NodeTransactionPoolService.ts:106]:
```suggestion
		const { maxBlockWeight } = this.api.consts.system.blockWeights;
```

File: src/services/node/NodeTransactionPoolService.ts [hunk 7]
```diff
@@ -14,24 +14,172 @@
...
+		const maxLength: u32 =
+			this.api.consts.system.blockLength.max[sanitizedClass];
+		const boundedWeight = BN.min(BN.max(weight.toBn(), BN_ONE), maxBlock);
+		const boundedLength = BN.min(BN.max(new BN(len), BN_ONE), maxLength);
+		const maxTxPerBlockWeight = maxBlock.div(boundedWeight);
+		const maxTxPerBlockLength = maxLength.div(boundedLength);
+
+		const maxTxPerBlock = BN.min(maxTxPerBlockWeight, maxTxPerBlockLength);
+		const saturatedTip = tip.toBn().add(BN_ONE);
+		const scaledTip = this.maxReward(saturatedTip, maxTxPerBlock);
```
Comment (niklasad1) [inline src/services/node/NodeTransactionPoolService.ts:116]: nice, totally missed that that was a closure in substrate

Comment (niklasad1) [inline src/services/node/NodeTransactionPoolService.ts:116]: nice, totally missed that that was a closure in substrate

**Review by niklasad1 — APPROVED** [review]
gucci :)

File: src/services/node/NodeTransactionPoolService.ts (hunk 6 above)
Comment (TarikGul) [reply src/services/node/NodeTransactionPoolService.ts:106]: So we can't actually rename this using this syntax as it's destructuring the key `maxBlock` from the object `blockWeights`, but I do agree that some explicit naming would be better here so I'll give it a change.

Comment (TarikGul) [reply src/services/node/NodeTransactionPoolService.ts]: changed

**Pull request merged**